spec/specification.md

```diff
@@ -0,0 +1,639 @@
+```
+C_1 |- e --> C_2
+```
+which is read as "under context `C_1`, `e` *steps to* context `C_2`".
+
+```
+C |-_t e : typ
+-------------------------------------------------------- # S-type-decl
+C |- (TypeDecl x e) --> C + symbols with x -> (type typ)
+
+x_1 notin C.symbols  x_1 != x_2 ...  x_2 notin C.symbols  |x_2| = |{x_2 ...}|  C |-_t e_1 : typ_1
+C |- e_2 : typ_2 ...  typ_2 != void ...  p = (ProcTy typ_1 typ_2 ...)
+C + return = typ + symbols with x_1 -> p, x_2 -> typ_2 ... |- e : void
+---------------------------------------------------------------------------------------------------- # S-proc-decl
+C |- (ProcDecl x_1 e_1 (Params (ParamDecl x_2 e_2)*) e_3) --> C + symbols with x_1 -> p
+
+
+-------------------- # S-empty-module
+C |- (Module) --> C
+
+C_n |- decl --> C_n+1 ...
+------------------------------- # S-module
+C_1 |- (Module decl+) --> C_n+1
+```
+
+A module is well-formed if it's either empty or all declarations within are
+well-formed.
+
+> TODO: the concept of an "entry procedure" is missing
+
+### Dynamic Semantics
+
+This section describes how expressions are evaluated, using small-step
+operational semantics with evaluation contexts (also known as *reduction semantics*
+or *context semantics*).
+
+#### Notation and Concepts
+
+Small-step operational semantics describe operation of a program as a series
+of *small steps* (hence the name).
+
+```
+configuration_1 --> configuration_2
+```
+is read as: "`configuration_1` *steps to* `configuration_2`".
+
+```
+premise_1  premise_2
+--------------------
+      a --> b
+```
+is read as: "if `premise_1` and `premise_2` hold true, it follows that `a`
+steps to `b`".
+
+A *notion of reduction* describes how a *redex* (reduction expression)
+reduces to a term. They're written as:
+```
+redex  ~~>  term
+```
+which is read as: "`redex` reduces to `term`".
+
+An *evaluation context* (usually named `E`) specifies where reduction may take
+place, and thus the *order of evaluation*. It represents an expression with
+a *hole* in it, with the hole denoted by `[]`. The *plug function*
+`E[t_1] = t_2` fills the hole of `E` with `t_1` yielding a new term `t_2`.
+
+Stepping is combined with the notions of reduction through an evaluation
+context:
+```
+   t_1 ~~> t_2
+-----------------
+E[t_1] --> E[t_2]
+```
+which is read as: "If `t_1` reduces to `t_2`, then `t_1` plugged into `E`
+steps to `t_2` plugged into `E`".
+
+#### Evaluation
+
+The configuration for steps is a tuple of the form `C; e`, where `C` is the
+*context* record. It's abstract syntax is as follows:
+```
+C ::= { locs S }
+```
+
+`locs` is a store (`S`), which maps *locations* to *values* (`S(l) = val`).
+Mutable state is modeled via locations.
+
+`e[x/y]` means "`e` with all occurrences of name `x` replaced with `y`".
+
+It is assumed that all identifiers referring to procedures were replaced with
+`(proc r [x p]* body)` prior to evaluation, where `x` and `typ` are the names
```

Comment:
```suggestion
`(proc r [x p]* body)` prior to evaluation, where `x` and `p` are the names
```

since the hole has `p` and not `typ`

Reply:
Oh, hm, the `[...]` is not intended to mean "hole" in this context (`typ` should be `p` nonetheless). It's only meant to group `x` an `p` together  (to apply the repetition `*` to) -- I didn't want to use parenthesis, as those are (usually) used for syntax tree nodes, hence the brackets.